odc/geo/_xr_interop.py

```diff
@@ -245,6 +245,59 @@ def assign_crs(
     return xx
 
 
+def crop(xx: XrT, poly: Geometry, mask: bool = True, all_touched: bool = True) -> XrT:
+    """
+    Crops and optionally mask an xr.Dataset or xr.DataArray (``xx``) to
+    the spatial extent of a geometry (``poly``).
+
+    :param xx:
+       :py:class:`~xarray.Dataset` or :py:class:`~xarray.DataArray`.
+
+    :param poly:
+       Geometry shape used to crop ``xx``.
+
+    :param poly:
+       Whether to mask out pixels outside of the rasterized extent of
+       ``poly``.
+
+    :param all_touched:
+        If ``True``, all pixels touched by ``poly`` will remain unmasked.
+        If ``False``, only pixels whose center is within the polygon or
+        that are selected by Bresenham's line algorithm will remain unmasked.
+
+    :return:
+        A :py:class:`~xarray.Dataset` or :py:class:`~xarray.DataArray`
+        cropped and optionally masked to the spatial extent of ``poly``.
+    """
+    # Reproject `poly` to match `xx` and calculate bounding box
+    poly = poly.to_crs(crs=xx.odc.crs)
+    bbox = poly.boundingbox
+
+    # Verify that `poly` overlaps with `xx` extent
+    if poly.intersects(xx.odc.geobox.extent):
+        # Crop `xx` to the bounding box of `poly`
+        ydim, xdim = spatial_dims(xx)
```

Comment:
mypy complains here about not handling the case when spatial_dims returns None.

Reply:
What would you recommend for that case? Just raise an exception? I 'm trying to think of a scenario where we'd expect this functionality to work with undefined spatial dims...

Reply:
assert is enough for mypy usualy, like

```python
sdims = spatial_dims(xx)
assert sdims
ydim, xdim = sdims
```

but here we should probably use `ydim` property that returns an integer index of `y` dimension, instead of dimension names.